packages/substrate-bindings/src/codecs/scale/Binary.ts

```diff
@@ -1,38 +1,69 @@
-import { Bytes, Codec, Decoder, Encoder, createCodec } from "scale-ts"
-import { fromHex, toHex } from "@polkadot-api/utils"
-import { HexString } from "./Hex"
+import {
+  Bytes,
+  Codec,
+  Decoder,
+  Encoder,
+  Tuple,
+  compact,
+  createCodec,
+} from "scale-ts"
+import { fromHex, mergeUint8, toHex } from "@polkadot-api/utils"
+import type { HexString } from "./Hex"
 
 const textEncoder = new TextEncoder()
 const textDecoder = new TextDecoder()
+const opaqueBytesDec = Tuple(compact, Bytes(Infinity))[1]
 
 export class Binary {
   #bytes: Uint8Array
+  #opaqueBytes: Uint8Array | null = null
   #hex: HexString | null = null
+  #opaqueHex: HexString | null = null
   #str: string | null = null
 
-  constructor(data: Uint8Array) {
-    this.#bytes = data
+  constructor(data: Uint8Array, opaque = false) {
+    if (opaque) {
+      try {
+        const [len, bytes] = opaqueBytesDec(data)
+        if (len === bytes.length) {
+          this.#bytes = bytes
+          this.#opaqueBytes = data
+          return
+        }
+      } catch (_) {}
+      throw new Error("Invalid opaque bytes")
+    } else this.#bytes = data
   }
 
-  asText = () =>
-    this.#str === null
-      ? (this.#str = textDecoder.decode(this.#bytes))
-      : this.#str
+  asText = () => (this.#str ||= textDecoder.decode(this.#bytes))
```

Comment:
I'd say it's safer to use `??=` instead. If the array is empty it defeats its purpose, for example.
Idem for the rest of the file.

Reply:
yep, good point!